File: common/src/main/java/project/price_it/security/JwtTokenProvider.java
package project.price_it.security;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;


import javax.crypto.SecretKey;
import java.util.Date;

@Component
public class JwtTokenProvider {
    private static final String AUTH_HEADER = "Authorization";
    private static final String BEARER_PREFIX = "Bearer ";

    @Value("${jwt.secret}")
    private String secretKey;

    @Value("${jwt.access-token-validity-ms}")
    private long accessTokenValidity;

    @Value("${jwt.refresh-token-validity-ms}")
    private long refreshTokenValidity;

    public String createAccessToken(String userId) {
        Date now = new Date();
        Date expiry = new Date(now.getTime() + accessTokenValidity);

        return Jwts.builder()
                .setSubject(userId)
                .setIssuedAt(now)
                .setExpiration(expiry)
                .signWith(SignatureAlgorithm.HS256, secretKey.getBytes())
                .compact();
    }

    public String createRefreshToken(String userId) {
        Date now = new Date();
        Date expiry = new Date(now.getTime() + refreshTokenValidity);

        SecretKey key = Keys.hmacShaKeyFor(secretKey.getBytes());

        return Jwts.builder()
                .setSubject(userId)
                .setIssuedAt(now)
                .setExpiration(expiry)
                .signWith(key, SignatureAlgorithm.HS256)
                .compact();
    }

    public String refreshAccessToken(String refreshToken) {
        try {
            Claims claims = Jwts.parser()
                    .setSigningKey(secretKey.getBytes())
                    .parseClaimsJws(refreshToken)
                    .getBody();

            String userId = claims.getSubject();
            return createAccessToken(userId);
        } catch (Exception e) {
            throw new IllegalArgumentException("토큰 유효하지 않음", e);
        }
    }


    public Long getUserIdFromAccessToken(String accessToken) {
        try {
            Claims claims = Jwts.parser()
                    .setSigningKey(secretKey.getBytes())
                    .parseClaimsJws(accessToken)
                    .getBody();

            String subject = claims.getSubject();
            return Long.parseLong(subject);
        } catch (JwtException | IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid JWT token", e);
        }
    }

    // 요청에서 JWT 추출
    public String resolveToken(HttpServletRequest request) {
        String bearerToken = request.getHeader(AUTH_HEADER);
        if (bearerToken != null && bearerToken.startsWith(BEARER_PREFIX)) {
            return bearerToken.substring(BEARER_PREFIX.length());
        }
        return null;
    }

}

File: common/src/main/java/project/price_it/security/JwtTokenProvider.kt
package project.price_it.security

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.Date
import javax.crypto.SecretKey

@Component
class JwtTokenProvider(
        @Value("\${jwt.secret}") private val secretKey: String,
        @Value("\${jwt.access-token-validity-ms}") private val accessTokenValidity: Long,
        @Value("\${jwt.refresh-token-validity-ms}") private val refreshTokenValidity: Long
) {

    fun createAccessToken(userId: String): String {
        val now = Date()
        val expiry = Date(now.time + accessTokenValidity)

        return Jwts.builder()
                .setSubject(userId)
                .setIssuedAt(now)
                .setExpiration(expiry)
                .signWith(SignatureAlgorithm.HS256, secretKey.toByteArray())
                .compact()
    }

    fun createRefreshToken(userId: String): String {
        val now = Date()
        val expiry = Date(now.time + refreshTokenValidity)
        val key: SecretKey = Keys.hmacShaKeyFor(secretKey.toByteArray())

        return Jwts.builder()
                .setSubject(userId)
                .setIssuedAt(now)
                .setExpiration(expiry)
                .signWith(key, SignatureAlgorithm.HS256)
                .compact()
    }

    fun refreshAccessToken(refreshToken: String): String {
        try {
            val claims: Claims = Jwts.parser()
                    .setSigningKey(secretKey.toByteArray())
                    .parseClaimsJws(refreshToken)
                    .body

            val userId = claims.subject
            return createAccessToken(userId)
        } catch (e: Exception) {
            throw IllegalArgumentException("토큰 유효하지 않음", e)
        }
    }

    fun getUserIdFromAccessToken(accessToken: String): String {
        val claims: Claims = Jwts.parser()
                .setSigningKey(secretKey.toByteArray())
                .parseClaimsJws(accessToken)
                .body

        return claims.subject
    }
}
